libdnf/sack/query.cpp

```diff
@@ -43,6 +43,98 @@
 
 namespace libdnf {
 
+struct NevraID {
+    Id name;
+    Id arch;
+    Id evr;
+    /**
+    * @brief Parsing function for nevra string into name, evr, arch and transorming it into libsolv
+    * Id
+    *
+    * @return bool Returns true if parsing succesfull and all elements is known to pool
+    */
+    bool parse(Pool * pool, const char * nevraPattern);
+};
+
+bool
+NevraID::parse(Pool * pool, const char * nevraPattern)
```

Comment:
I tried to create a shorter, better optimized and more readable code.
My version has advantages:
1. It uses the `pool_strn2id()` procedure (your solution uses `pool_str2id()` which internaly call `strlen()` and then `pool_strn2id()`).
2. It does not modify parsed string, so no need to make copy of input string.
3. There is only one pass throught parsed string (no additional passes hidden in `strlen()` and `strcpy()`)

```
bool
NevraID::parse(Pool * pool, const char * nevraPattern)
{
    const char * evrDelim = nullptr;
    const char * releaseDelim = nullptr;
    const char * archDelim = nullptr;
    const char * end;

    // parse nevra
    for (end = nevraPattern; *end != '\0'; ++end) {
        if (*end == '-') {
            evrDelim = releaseDelim;
            releaseDelim = end;
        } else if (*end == '.') {
            archDelim = end;
        }
    }

    // test name presence
    if (!evrDelim || evrDelim == nevraPattern)
        return false;

    auto nameLen = evrDelim - nevraPattern;

    // strip epoch "0:"
    if (evrDelim[1] == '0' && evrDelim[2] == ':')
        evrDelim += 2;

    // test version and arch presence
    if (releaseDelim - evrDelim <= 1 ||
        !archDelim || archDelim <= releaseDelim + 1 || archDelim == end - 1)
        return false;


    // convert strings to Ids
    if (!(name = pool_strn2id(pool, nevraPattern, nameLen, 0)))
        return false;
    ++evrDelim;
    if (!(evr = pool_strn2id(pool, evrDelim, archDelim - evrDelim, 0)))
        return false;
    ++archDelim;
    if (!(arch = pool_strn2id(pool, archDelim, end - archDelim, 0)))
        return false;

    return true;
}
```

Reply:
This is still not best algorithm, although is pretty good ;)

You better iterate string from right to left because in case you have very long name, you waste cycles... Not that it matters much here, but... just sayin :)